clang-tools-extra/clang-tidy/hooks/GuardInForCheck.cpp

```diff
@@ -19,116 +19,240 @@ namespace hooks {
 namespace {
 
 const int64_t MAX_FOR_LIMIT = 10000;
+const int MAX_NESTED_LOOP_LEVEL = 10;
 
 bool areSameVariable(const ValueDecl *First, const ValueDecl *Second) {
   return First && Second &&
     First->getCanonicalDecl() == Second->getCanonicalDecl();
 }
 
+auto isUnguarded()  {
+  return unless(has(binaryOperator(has(callExpr(callee(functionDecl(hasName("_g"))))))));
 }
 
-void GuardInForCheck::registerMatchers(MatchFinder *Finder) {
-  // based on https://clang.llvm.org/docs/LibASTMatchersTutorial.html
-  const auto CondVarNameExpr = ignoringParenImpCasts(declRefExpr(
-          to(varDecl(hasType(isInteger())).bind("condVarName"))));
+auto conditionLimit(int nestingLevel) {
+  return expr().bind("constLimit-" + std::to_string(nestingLevel));
+}
+
+auto conditionVarName(int nestingLevel) {
+  return ignoringParenImpCasts(declRefExpr(
+          to(varDecl(hasType(isInteger())).bind("condVarName-" + std::to_string(nestingLevel)))));
+}
 
-  const auto CondLimitExpr = expr().bind("constLimit");
+auto lessThanCondition(int nestingLevel) {
+  return binaryOperator(hasOperatorName("<"),    //e.g. i < 2
+            hasLHS(conditionVarName(nestingLevel)),
+            hasRHS(conditionLimit(nestingLevel))).bind("condOp-" + std::to_string(nestingLevel));
+}
+
+auto greaterThanCondition(int nestingLevel) {
+  return binaryOperator(hasOperatorName(">"),    //e.g. 2 > i
+            hasLHS(conditionLimit(nestingLevel)),
+            hasRHS(conditionVarName(nestingLevel))).bind("condOp-" + std::to_string(nestingLevel));
+}
+
+auto loopCondition(int nestingLevel) {
+  return hasCondition(
+        anyOf(
+          binaryOperator(hasOperatorName(","),    //catch loop that is guarded, this is for nested loops only - even if one of ancestor loops is guarded,
+                                                  //it's condition limit still need to be used to calculate GUARD limit for the most descendant loop
```

Comment:
I'm confused by the comment... :-) Shouldn't it say that for nested loops, calculating the guard limit needs the limits of all its ancestors?

Reply:
That specific comment is for cases where one of the ancestor loops is guarded - in such a case we should not ignore it, but get its limit (together with limits of all other ancestors) to calculate the limit of most descendant loop. I'll refactor it to make it more clear

Reply:
you mean the case like
```
    for (int i = 0; GUARD(3), i != 3; ++i)
	     for (int j = 0; j < 4; ++j)
		    trace_num("j", 1, j);
```
where the inner loop gets the outer limit from the argument to `_g`? That could indeed work, but currently it doesn't... :-)

Reply:
The limit itself is still taken from the RHS of the `<` condition (or LHS in case of something like `2 > i`)

Reply:
well, that's a different issue - I meant an example like
```
for (int i = 0; GUARD(3), i < (3 + reserved); ++i)
	     for (int j = 0; j < 4; ++j)
		    trace_num("j", 1, j);
```
where the programmer knows what the limit is, _and said so_, even if the compiler doesn't...

Reply:
Ok, I get it. Still I'll need to fix it to take the limit out of the GUARD argument.

Reply:
Well, we don't have to implement _everything_ that crosses my mind... :-) There'll come a point where the template expansion gets too much for LLVM, or the resulting matcher too slow - we have to prioritize...

Reply:
OK, the cases above now work (the fix is offered), but not for
```
    uint64_t j = 10000000000000000000ULL;
    uint64_t unsigned_num = 42;
    for (; GUARD(20), unsigned_num > 0 && j > 0; j /= 10)
    {
        for (int k = 0; k < 3; ++k)
	        trace_num("k", 1, k);
	}
```
Why? Isn't it the same case?

Reply:
Currently simple for loops are supported, here 3 things are not handled:
 - missing expression in init part of the loop 
 - too big initial value of j
 - the for increment part - currently only ++ and -- are supported
 
 I can make the fixes, but aren't we going out of the task scope a little bit here? The initial task was to handle properly nested loops, not all kind of different for loop forms (and I'm sure it's not possible to handle all of them anyway).

Reply:
Well, yes, we are getting out of scope - as I said, I do not insist on implementing things just because we're discussing them... It's just that before check documentation update, I'd like to know what the new behavior is (and should be)... Also, the fix cannot be deployed until we have the staging server anyway...

Reply:
Ok, maybe it would be good to write down somewhere what cases should be handled, instead of pointing them out during review and going back to implementation :)

Reply:
Well, that would have the risk of specifying features which are not, in fact, reasonable to implement... :-) On a high level, `clangd` is just a linter; even if it wasn't there at all, the IDE would still be usable - so it doesn't have any really hard requirements, and the challenge is to come up with a set of helpful cases, which IMHO cannot be done before implementing (some of) them... But of course we (you :-) ) can write down the spec - how would you update https://github.com/XRPLF/xrpl-hooks-ide/blob/main/xrpl-hooks-docs/md/hooks-guard-in-for.md ?

Reply:
Ok, let me think about it and I'll try to prepare something there.

Reply:
Actually thinking about it, the user-facing docs are just a summary of a comment in source code, in this case in `GuardInForCheck.h` - that's what should be updated first.